File: neetcode-kotlin/app/src/main/kotlin/com/niyaj/neetcode/ReverseLinkedList.kt
package com.niyaj.neetcode


/**
 * Reverse Linked List
 * Given the beginning of a singly linked list head, reverse the list, and return the new beginning of the list.
 *
 * Example 1:
 *
 * Input: head = [0,1,2,3]
 *
 * Output: [3,2,1,0]
 * Example 2:
 *
 * Input: head = []
 *
 * Output: []
 * Constraints:
 *
 * 0 <= The length of the list <= 1000.
 * -1000 <= Node.val <= 1000
 */
private class ReverseLinkedList constructor(var value: Int = -1, var next: ReverseLinkedList? = null) {
    override fun toString(): String {
        return "$value -> ${next.toString()}"
    }

    companion object {
        // a help function to generate a linked list with given values quickly, for test purpose only
        fun quickList(nodes: List<Int>): ReverseLinkedList {
            val dummy = ReverseLinkedList()
            nodes.reversed().forEach({
                val temp = ReverseLinkedList(it)
                temp.next = dummy.next
                dummy.next = temp
            })
            return dummy.next!!
        }
    }
}

fun main() {
    val listNode = ReverseLinkedList.quickList(listOf(0, 1, 2, 3))
    println(listNode)
}

File: neetcode-cpp/ReverseLinkedList.cpp
#include <iostream>

#include "ListNode.h"

using namespace std;

class ReverseLinkedList {
public:
    static ListNode *reverseList(ListNode *head) {
        ListNode *prev = nullptr;
        ListNode *curr = head;

        while (curr) {
            ListNode *temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }

        return prev;
    }
};

int main() {
    // Create a linked list: 1 -> 2 -> 3 -> 4 -> 5
    auto* head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));

    std::cout << "Original List: ";
    ListNode::printList(head);

    // Reverse the linked list
    ListNode* reversedHead = ReverseLinkedList::reverseList(head);

    std::cout << "Reversed List: ";
    ListNode::printList(reversedHead);

    return 0;
}
